File: q26-50/q46.py
# 2020.08.17
# Happy birthday! (not to me)
# Problem Statement:
# https://leetcode.com/problems/permutations/

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # check corner cases and do early return
        if len(nums) == 0: return [[]]       
        if len(nums) == 1: return [[nums[0]]]

        # idea:
        # from just 2 element (2 permutation arrays), 
        # insert the 3rd elememnt to all possible positions and make it 6 arrays in total;
        # then insert the 4th element to the 6 arrays with 3 elements, 
        # do all the way until all number in nums are added.

        # initialize a data structure to hold the initial answer
        answer = [[nums[0], nums[1]], [nums[1], nums[0]]]
        
        # early return
        if len(nums) == 2: return answer

        # always do append in new_answer, and update answer to new_answer, and make new_answer empty again
        new_answer = []

        # add all the rest elements
        for num in nums[2:]:
            # add to each existing array
            for array in answer:  
                # iterate through all the possible positions the new number can insert into              
                for i in range(0, len(array)+1):           
                    new_answer.append(array[:i] + [num] + array[i:])  
            # have to do deep copy, instead of shallow copy
            answer = deepcopy(new_answer)
            # clear up new_answer
            new_answer = []
    
        return answer

File: Amazon/q46.java
// 2022.06.12
// Problem Statement:
// https://leetcode.com/problems/permutations/

// idea: recursion, save the first element and insert it into all possible positions
// in the result of the subarray (excluding the first element)
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> answer = new ArrayList<> ();
        // base case 1
        if (nums.length==0) {
            List<Integer> one_answer = new ArrayList<> ();
            answer.add(one_answer);
            return answer;
        }
        // base case 2
        if (nums.length==1) {
            List<Integer> one_answer = new ArrayList<> ();
            one_answer.add(nums[0]);
            answer.add(one_answer);
            return answer;
        }
        List<List<Integer>> rest = permute(Arrays.copyOfRange(nums, 1, nums.length));
        for (List<Integer> l : rest) {
            for (int i=0; i<=l.size(); i++) {
                List<Integer> copy = new ArrayList<> (l);
                copy.add(i, nums[0]);
                answer.add(copy);
            }
        }
        return answer;
    }
}